File: Go/FindTheClosestPalindrome.go
//go:build ignore

package main

import (
	"fmt"
	"strconv"
)

func reverse(s string) string {
	b := []byte(s)
	for i, j := 0, len(b)-1; i < j; i, j = i+1, j-1 {
		b[i], b[j] = b[j], b[i]
	}
	return string(b)
}

func abs64(x int64) int64 {
	if x < 0 {
		return -x
	}
	return x
}

func nearestPalindromic(n string) string {
	length := len(n)
	num, _ := strconv.ParseInt(n, 10, 64)
	cands := map[int64]bool{}
	cands[pow10(length-1)-1] = true
	cands[pow10(length)+1] = true

	prefix, _ := strconv.ParseInt(n[:(length+1)/2], 10, 64)
	for _, p := range []int64{prefix - 1, prefix, prefix + 1} {
		s := strconv.FormatInt(p, 10)
		var cand string
		if length%2 == 0 {
			cand = s + reverse(s)
		} else {
			cand = s + reverse(s[:len(s)-1])
		}
		v, _ := strconv.ParseInt(cand, 10, 64)
		cands[v] = true
	}
	delete(cands, num)

	var best int64 = -1
	for c := range cands {
		if c < 0 {
			continue
		}
		if best == -1 || abs64(c-num) < abs64(best-num) ||
			(abs64(c-num) == abs64(best-num) && c < best) {
			best = c
		}
	}
	return strconv.FormatInt(best, 10)
}

func pow10(e int) int64 {
	r := int64(1)
	for range e {
		r *= 10
	}
	return r
}

func main() {
	for _, t := range []string{"123", "1", "10", "1000", "999"} {
		fmt.Println(nearestPalindromic(t)) // 121, 0, 9, 999, 1001
	}
}

File: CPP/FindTheClosestPalindrome.cpp
#include <cstdlib>
#include <iostream>
#include <set>
#include <string>

class Solution {
public:
  std::string nearestPalindromic(std::string n) {
    int length = (int)n.size();
    long long num = std::stoll(n);
    std::set<long long> cands;
    cands.insert(pow10(length - 1) - 1);
    cands.insert(pow10(length) + 1);

    long long prefix = std::stoll(n.substr(0, (length + 1) / 2));
    for (long long p : {prefix - 1, prefix, prefix + 1}) {
      std::string s = std::to_string(p);
      std::string rev(s.rbegin(), s.rend());
      std::string cand = (length % 2 == 0) ? s + rev : s + rev.substr(1);
      cands.insert(std::stoll(cand));
    }
    cands.erase(num);

    long long best = -1;
    for (long long c : cands) {
      if (c < 0)
        continue;
      if (best == -1 || llabs(c - num) < llabs(best - num) ||
          (llabs(c - num) == llabs(best - num) && c < best)) {
        best = c;
      }
    }
    return std::to_string(best);
  }

private:
  long long pow10(int e) {
    long long r = 1;
    for (int i = 0; i < e; ++i)
      r *= 10;
    return r;
  }
};

int main() {
  Solution s;
  for (std::string t : {"123", "1", "10", "1000", "999"}) {
    std::cout << s.nearestPalindromic(t) << "\n"; // 121, 0, 9, 999, 1001
  }
  return 0;
}
